src/Serializer/DefaultHydrator.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Patchlevel\EventSourcing\Serializer;
+
+use Patchlevel\EventSourcing\Attribute\Normalize;
+use ReflectionClass;
+use ReflectionProperty;
+use RuntimeException;
+use TypeError;
+
+use function array_key_exists;
+use function assert;
+use function is_string;
+
+final class DefaultHydrator implements Hydrator
+{
+    /** @var array<class-string, array<string, array{reflection: ReflectionProperty, fieldName: string, normalizer: ?Normalizer}>> */
+    private static array $propertyMetadataCache = [];
+
+    /** @var array<class-string, ReflectionClass> */
+    private static array $reflectionClassCache = [];
+
+    /**
+     * @param class-string<T>      $class
+     * @param array<string, mixed> $data
+     *
+     * @return T
+     *
+     * @template T of object
+     */
+    public function hydrate(string $class, array $data): object
+    {
+        $object = self::reflectionClass($class)->newInstanceWithoutConstructor();
+        assert($object instanceof $class);
```

Comment:
would say: lets take an if + exception here instead of assert

Reply:
This is a case that can never happen. It's only for the psalm, because otherwise he wouldn't understand it. I don't really want to do a check, so assert fits better since it's ignored in production.